File: MaximumLengthSnakeSequence.java
package go;

public class MaximumLengthSnakeSequence {

	public static int maximumLengthSnakeSequenceSolution(int[][] matrix) {

		int maxSnakeLength = 0;
		int rows = matrix.length;
		int cols = matrix[0].length;
		int pivot = cols - 1; // 0 index ready

		int[][] dp = new int[rows][cols];
		dp[rows - 1][cols - 1] = 1;

		for (int pi = pivot - 1; pi >= 0; pi--) {
			int piElement = dp[pi][pi];
			for (int j = rows - 1; j > pi; j--) {

				int element = matrix[j][pivot];
				int elementRight = matrix[j][pi + 1];
				int elementBottom = (j != rows - 1 ? matrix[j + 1][pi] : 0);

				if (Math.abs(element - elementRight) == 1
						|| (j != rows - 1 && Math.abs(element - elementBottom) == 1)) {
					dp[j][pi] = Math.max(dp[j][pi + 1], (j != rows - 1 ? dp[j + 1][pi] : 0))+1;
					maxSnakeLength = Math.max(maxSnakeLength, dp[j][pi]);
				}

			}
			for (int i = cols - 1; i >= pivot; i--) {

				int element = matrix[pi][i];
				int elementRight = (i != cols - 1 ? matrix[pi][i + 1] : 0);
				int elementBottom = matrix[pi + 1][i];

				if (Math.abs(elementBottom - element) == 1
						|| (i != cols - 1 && Math.abs(elementRight - element) == 1)) {

					dp[pi][i] = Math.max(dp[pi + 1][i], (i != cols - 1 ? dp[pi][i + 1] : 0))+1;
					maxSnakeLength = Math.max(maxSnakeLength, dp[pi][i]);
				}
			}

		}

		return maxSnakeLength;

	}

	public static void main(String args[]) {

		int[][] matrix = new int[][] { { 9, 6, 5, 2 }, { 8, 7, 6, 5 }, { 7, 3, 1, 6 }, { 1, 1, 1, 7 }, };
		System.out.println(maximumLengthSnakeSequenceSolution(matrix));

	}
}

File: MaximumLengthSnakeSequence.py
"""
Find maximum length Snake sequence
Given a grid of numbers, find maximum length Snake sequence and print it. If multiple snake sequences exists with the maximum length, print any one of them.

A snake sequence is made up of adjacent numbers in the grid such that for each number, the number on the right or the number below it is +1 or -1 its value. For example, if you are at location (x, y) in the grid, you can either move right i.e. (x, y+1) if that number is ± 1 or move down i.e. (x+1, y) if that number is ± 1.

For example,

9, 6, 5, 2
8, 7, 6, 5
7, 3, 1, 6
1, 1, 1, 7

In above grid, the longest snake sequence is: (9, 8, 7, 6, 5, 6, 7)

Below figure shows all possible paths –

LINK: https://www.geeksforgeeks.org/find-maximum-length-snake-sequence/
Temporal maker: 17:04 Hours | Aug18, Sunday
Temporal Marker Untethered: 17:34 Hours | Aug18, Sunday
"""

#A utility function that prints a matrix
#Paramter: a 2D Matrix
#Returns: None
def printMatrix(matrix):
    for x in matrix:
        for y in x:
            print(str(y)+" ", end=" ")
        print()


#A utility function that finds maximum length Snake sequence and print it
#Paramter: a 2D Matrix
#Returns: Maximum Lengthed Snake Sequence
def findMaximumLengthSnakeSequence(array):
    rows = len(array)
    cols = len(array[0])
    snakeLength = 0

    dp = [[1]*(cols) for x in range(rows)]
    #printMatrix(dp)
                    
    for i in range(rows-1, -1, -1):
        for j in range(cols-1, -1, -1):
            #print("i: "+str(i)+" j: "+str(j))
            branchRight = 0
            branchBottom = 0
            if(j<cols-1 and abs(array[i][j+1]-array[i][j])==1):
                branchRight = dp[i][j+1]+1
            if(i<rows-1 and abs(array[i][j]-array[i+1][j])==1):
                branchBottom = dp[i+1][j]+1
            dp[i][j] = max(dp[i][j], branchRight, branchBottom)
            snakeLength = max(snakeLength, dp[i][j])
            #printMatrix(dp)

    #Returns the answer
    return snakeLength
       

grid =[[9, 6, 5, 2],[
8, 7, 6, 5],[
7, 3, 1, 6],[
1, 1, 1, 7]]
        
print(findMaximumLengthSnakeSequence(grid))
